From hsggebhardt@googlemail.com  Thu Nov  5 10:08:03 2009
From: Henry Gebhardt <hsggebhardt@googlemail.com>
Date: Wed, 4 Nov 2009 11:19:28 +0100
Subject: USB: cdc_acm: Fix race condition when opening tty
To: Oliver Neukum <oliver@neukum.org>
Cc: greg@kroah.com
Message-ID: <1bd893420911040219vbfe58a1redb8c8ed934760bd@mail.gmail.com>

From: Henry Gebhardt <gebhardt@astro.uni-tuebingen.de>

If acm_rx_tasklet() gets called before tty_port_block_til_ready()
returns, then bulk IN urbs may not be sent. This fixes it.

Signed-off-by: Henry Gebhardt <gebhardt@astro.uni-tuebingen.de>
Acked-by: Oliver Neukum <oliver@neukum.org>
Cc: stable <stable@kernel.org>
Signed-off-by: Greg Kroah-Hartman <gregkh@suse.de>

---
 drivers/usb/class/cdc-acm.c |    2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

--- a/drivers/usb/class/cdc-acm.c
+++ b/drivers/usb/class/cdc-acm.c
@@ -609,9 +609,9 @@ static int acm_tty_open(struct tty_struc
 
 	acm->throttle = 0;
 
-	tasklet_schedule(&acm->urb_task);
 	set_bit(ASYNCB_INITIALIZED, &acm->port.flags);
 	rv = tty_port_block_til_ready(&acm->port, tty, filp);
+	tasklet_schedule(&acm->urb_task);
 done:
 	mutex_unlock(&acm->mutex);
 err_out:
